javatests/arcs/core/storage/ReferenceModeStoreDatabaseIntegrationTest.kt

```diff
@@ -0,0 +1,546 @@
+        val activeStore = createReferenceModeStore()
+        val actor = activeStore.crdtKey
+
+        val personCollection = CrdtSet<RawEntity>()
+        val bob = createPersonEntity("an-id", "bob", 42)
+        val operation = CrdtSet.Operation.Add("me", VersionMap("me" to 1), bob)
+
+        val referenceCollection = CrdtSet<Reference>()
+        val bobRef = Reference(
+            "an-id",
+            activeStore.backingStore.storageKey,
+            VersionMap(actor to 1)
+        )
+        val refOperation = CrdtSet.Operation.Add(actor, VersionMap(actor to 1), bobRef)
+
+        val bobEntity = createPersonEntityCrdt()
+
+        // Apply to RefMode store.
+        assertThat(
+            activeStore.onProxyMessage(
+                ProxyMessage.Operations(
+                    listOf(RefModeStoreOp.SetAdd(actor, VersionMap(actor to 1), bob)),
+                    id = 1
+                )
+            )
+        ).isTrue()
+
+        // Apply to expected collection representation
+        assertThat(personCollection.applyOperation(operation)).isTrue()
+        // Apply to expected refMode collectionStore data.
+        assertThat(referenceCollection.applyOperation(refOperation)).isTrue()
+        // Apply to expected refMode backingStore data.
+        assertThat(
+            bobEntity.applyOperation(
+                CrdtEntity.Operation.SetSingleton(
+                    actor,
+                    VersionMap(actor to 1),
+                    "name",
+                    CrdtEntity.ReferenceImpl("bob".toReferencable().id)
+                )
+            )
+        ).isTrue()
+        assertThat(
+            bobEntity.applyOperation(
+                CrdtEntity.Operation.SetSingleton(
+                    actor,
+                    VersionMap(actor to 1),
+                    "age",
+                    CrdtEntity.ReferenceImpl(42.toReferencable().id)
+                )
+            )
+        ).isTrue()
+
+        val containerKey = activeStore.containerStore.storageKey as DatabaseStorageKey
+        val capturedPeople =
+            databaseFactory.getDatabase(containerKey.dbName, containerKey.persistent)
+                .get(containerKey, DatabaseData.Collection::class) as DatabaseData.Collection
+
+        assertThat(capturedPeople.values).isEqualTo(referenceCollection.consumerView)
+        val storedBob = activeStore.backingStore.getLocalData("an-id") as CrdtEntity.Data
+        // Check that the stored bob's singleton data is equal to the expected bob's singleton data
+        assertThat(storedBob.singletons.mapValues { it.value.data })
+            .isEqualTo(bobEntity.data.singletons.mapValues { it.value.data })
+        // Check that the stored bob's collection data is equal to the expected bob's collection
+        // data (empty)
+        assertThat(storedBob.collections.mapValues { it.value.data })
+            .isEqualTo(bobEntity.data.collections.mapValues { it.value.data })
```

Comment:
I think you should just be able to do an equality check:

```
assertThat(storedBob.singletons).isEqualTo(bobEntity.data.singletons)
```

Reply:
Thanks!